### ISSUE TITLE
[bug] cmake generator does not write build requirements info when using dual profiles

### ISSUE THREAD

**Issue by theodelrieu** [conan-io/conan#8303]
### Environment Details (include every applicable attribute)
  * Operating System+version: macOS Big Sur
  * Conan version: develop (1.33.0-dev)
  * Python version: 3.8.5

### Steps to reproduce (Include if Applicable)

With this `conanfile.py`:

```python
from conans import ConanFile, CMake, tools


class BugConan(ConanFile):
    name = "bug"
    version = "0.1"
    license = "<Put the package license here>"
    author = "<Put your name here> <And your email here>"
    url = "<Package recipe repository url here, for issues about the package>"
    description = "<Description of Bug here>"
    topics = ("<Put some tag here>", "<here>", "<and here>")
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": False, "fPIC": True}
    generators = "cmake"

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def build_requirements(self):
        self.build_requires("doctest/2.3.8")

    def source(self):
        pass

    def build(self):
        pass

    def package(self):
        pass

    def package_info(self):
        pass
```

Run `conan install . --profile:host=default --profile:build=default`

The generated `conanbuildinfo.cmake` does not contain `doctest` targets, but it does when `conan install . --profile=default` is used.

**Comment by theodelrieu** [comment]
This is because of the line `all_flags = cmake_dependencies(dependencies=self.deps_build_info.deps)`.

`self.deps_build_info` is empty when using dual profiles

**Comment by theodelrieu** [comment]
I've added logs in `model/conan_generator.py:

With `--profile=default`:

>self._deps_build_info: <conans.model.env_info.DepsEnvInfo object at 0x105c68b50>
>self._deps_env_info <conans.model.env_info.DepsEnvInfo object at 0x105c68b50>
>self._env_info None
>self._deps_user_info DepsUserInfo(<class 'conans.model.user_info.UserInfo'>, {'doctest': {}})
>self._user_info_build None

With dual profiles:

>self._deps_build_info: <conans.model.env_info.DepsEnvInfo object at 0x1061e58b0>
>self._deps_env_info <conans.model.env_info.DepsEnvInfo object at 0x1061e58b0>
>self._env_info None
>self._deps_user_info DepsUserInfo(<class 'conans.model.user_info.UserInfo'>, {})
>self._user_info_build DepsUserInfo(<class 'conans.model.user_info.UserInfo'>, {'doctest': {}})

Seems that generators do not support `user_info_build`

**Comment by jgsogo** [comment]
Hi, @theodelrieu 

This is probably a limitation of the `cmake` generator. With the current syntax there is no way to differentiate between a package coming from _host_ context and the same package from _build_ context if both have the same name (i.e.: `protobuf`). It would require variables like `CONAN_USER_<PKG-NAME>_<VAR-NAME>` and `CONAN_USER_<CONTEXT>_<PKG-NAME>_<VAR-NAME>`.

Doable, but not sure if we want to go that way. If the preferred generator `cmake_find_package[_multi]` doesn't provide this functionality, IMHO we shouldn't add it to a generator that might be removed in Conan v2.0.... or at least we shouldn't add it before implementing it in the recommended generator.

But we totally need to unblock this issue, there are many related to this same problem or closer ones. I really think there is something missing in CMake itself, but we need to bet big and offer a way to work around current limitations. I'm thinking about opening an RFC with some risky proposal.

**Comment by theodelrieu** [comment]
> If the preferred generator cmake_find_package[_multi] doesn't provide this functionality

Indeed, the Finddoctest.cmake is only generated when `--profile default` is used.

I don't know how I can workaround this issue, I don't want to revert to `os_build/arch_build` but I might not have a choice here...

**Comment by madebr** [comment]
@theodelrieu 
Why do you add doctest as a build requirement?
doctest is a header-only library. Shouldn't it be a requirement?

**Comment by jgsogo** [comment]
Hi, @theodelrieu 

We've been talking about this issue (and related ones) and the path will be as follows: 
 * Conan will (1.34?) generate a file with all the information from `cpp_info` in plain CMake format. This file will contain the information as plain as possible. It will be hard to define the name of all the variables, but the file will contain information from all the packages in _host_ and _build_ context that are needed to build the package. (Note.- Maybe it is not only one file, but one per requirement, TBD).
 * This file will be accessible only if using generator `CMakeDeps` (in fact, this generator will use that file/s to populate the targets). Generator `CMakeDeps` will be the recommended (only?) one in Conan 2.0 and we encourage everyone to try it and start using it to smooth the migration.
 * User can `include` that file from their `CMakeLists.txt` and will have access to all these variables.
 * Eventually, Conan will provide some abstraction layers on top of this file to provide targets for the components coming from the _build_ context (protoc, doxygen,...). And some convenient CMake functions (`find_package_build`) or overrides (new definition of `find_package`) that will help with these scenarios and make it possible some transparent integration of Conan with existing CMake files.

**Comment by blackliner** [comment]
In my case I need [CapnProto ](https://conan.io/center/capnp?tab=overview) to be compiled for the build system, generate sources, and then be available as `find_package` for building the host binaries.

Is there any workaround for cross compilation right now?

EDIT: maybe a two step `conan install`, so instead of
`conan install .. --profile:build build_profile --profile:host host_profile`
we split it up and call
`conan install package_A package_B --profile build_profile`
--> copy away the files you need...
`conan install package_C package_D --profile host_profile`
--> ... and restore them again
but for that I would need to evaluate the `conanfile.py` manually I guess 😞

**Comment by blackliner** [comment]
As seen below, `--profile:build` just completely skips the generation of any `*.cmake` files for packages listed as `build_requirements`. For `cmake_paths`, `conan_paths.cmake` does not contain any info about them as well.

For this `conanfile.py`:
```python
    generators = "cmake_find_package"
    ...
    def build_requirements(self):
        self.build_requires("capnproto/0.8.0")

    def requirements(self):
        self.requires("catch2/2.13.3")
```

`conan install .. --profile:build default`:
```python
conanfile.py (LumPDK/None): Applying build-requirement: capnproto/0.8.0
conanfile.py (LumPDK/None): Generator cmake_find_package created FindCatch2.cmake
```

`conan install .. --profile:host default`:
```python
conanfile.py (LumPDK/None): Applying build-requirement: capnproto/0.8.0
conanfile.py (LumPDK/None): Generator cmake_find_package created FindCapnProto.cmake
conanfile.py (LumPDK/None): Generator cmake_find_package created FindCatch2.cmake
```

`conan install .. --profile default`:
```python
conanfile.py (LumPDK/None): Applying build-requirement: capnproto/0.8.0
conanfile.py (LumPDK/None): Generator cmake_find_package created FindCapnProto.cmake
conanfile.py (LumPDK/None): Generator cmake_find_package created FindCatch2.cmake
```

**Comment by jgsogo** [comment]
@blackliner The way to go is to add `capnproto` to requires and build-requires:

```python

    generators = "cmake_find_package"
    ...
    def build_requirements(self):
        self.build_requires("capnproto/0.8.0")

    def requirements(self):
        self.requires("capnproto/2.13.3")
        self.requires("catch2/2.13.3")

    def build(self):
        # All the environment provided by `capnproto` from the build context is available here.

```

and then use the two-profiles approach:

```
conan create conanfile.py --profile:host=host --profile:build=default
```

Conan will populate the environment before entering `build()` method with the information provided in the `env_info` of the `capnproto` recipe.

**Comment by blackliner** [comment]
Unfortunately we do not use conan to build our own package, we just use it to consume dependencies. I went for this solution/workaround for now:
```python
def conan(args):
    if not (BASE_DIR / "conanfile.py").is_file():
        logging.info("Skipping conan step")
        return

    export_recipes(args)

    conan_command = [
        "conan",
        "install",
        "..",
        "--settings",
        "build_type=" + args.build_type,
        "--build=missing",
        "--build=acado",
    ]

    if args.conan_profile:
        conan_command.append(f"--profile={CALLER_DIR / args.conan_profile}")
    else:        
        if args.conan_build_profile:
            conan_command_build_profile = conan_command.copy()
            conan_command_build_profile.append(f"--profile={CALLER_DIR / args.conan_build_profile}")
            subprocess.check_call(conan_command_build_profile, cwd=BUILD_DIR)
            for cmake_file in BUILD_DIR.glob("*.cmake"):
                cmake_file.rename(cmake_file.with_suffix(".build_bak"))

            conan_command.append(f"--profile:build={CALLER_DIR / args.conan_build_profile}")

        if args.conan_host_profile:
            conan_command.append(f"--profile:host={CALLER_DIR / args.conan_host_profile}")

    logging.info("Conan command:")
    logging.info(" ".join(conan_command))

    subprocess.check_call(conan_command, cwd=BUILD_DIR)

    for cmake_bak_file in BUILD_DIR.glob("*.build_bak"):
        new_name = cmake_bak_file.with_suffix(".cmake")
        if not new_name.is_file():
            cmake_bak_file.rename(new_name)
```
So for now I save all the `FindXXX.cmake` from the build context and restore them later 🤷

**Comment by Hopobcn** [comment]
What about using `force_host_context=True`?
In my case, I'm experimenting with `CMakeToolchain` and two profiles. And unfortunately, I saw this issue when providing the same profile for build & host contexts.

If I add `force_host_context=True`:
```
    def build_requirements(self):
        if self.options.get_safe("build_tests"):
            self.build_requires("gtest/1.10.0", force_host_context=True)
```
I'm able to compile & run tests in the `build` method.

**Comment by ericriff** [comment]
I just came across this limitation myself.
And it also applies to the `cmake_paths` generator.

On `conanbuildinfo.cmake` there is no reference to the build_requierements at all. Since we're also using the `virtualenv` generator, the `PATH` is fine, but we don't have access to anything else like `CMAKE_PREFIX_PATH` to use some cmake tools those packages provide.
Any suggested workaround?

EDIT: This information IS present on the `conanbuildinfo.txt`, but not on the `.cmake` version.

**Comment by SeanSnyders** [comment]
Yes, I also logged an issue on the same topic with a good repro using conan's own hello-world:
https://github.com/conan-io/conan/issues/8839

**Comment by lasote** [comment]
Hi all. We are taking a look again at this issue.

**To summarize the current status**
When the new cross-building mechanism (two profiles) is activated, the declared build_requires are in `build` context, and Conan is not generating files for the requirements in the `build` context.

Why? Well, there might be potential collisions when a package is in `build context` and in the `host context`. For example, `protobuf` package from conan-center might be included as require (because you want to link with the `protobuf` library) and also as a build_require, because you want to call `protoc` to generate the sources. Right? So far so good, but imagine you are cross-building from, let's say from Linux (build) to Windows (host). Now the `protoc` has to run in Linux but the protobuf library has to be compiled for Windows. So, there are two different `package_id` at the Conan cache... and which one should be pointing the `config` cmake file? There is no good answer for that.

**Proposal**
We think we might be able to generate the config files for the `build` context (for the new `CMakeDeps` generator only) but with a suffix for the file name, something like `ProtobufBuildConfig.cmake` or similar. It would require changing the consumer `find_package()` to specify the new name. This way we could keep packages like `protobuff` containing the `protoc` and the `libprotobuf` without collisions. 

It is the only way we think we can solve this without splitting the packages like protobuff into two packages, which was clearly refused by the community from the `conan-center`.

WDYT?

**Comment by ericriff** [comment]
We use protobuf as both a `requirement` and as a `build_requirement` for the reasons you're naming: We crosscompile from `Linux:x86_64` to `Linux:arm` so we need an x86_64 package for `protoc` and an arm version for the libraries.
The main issue we're facing is an internal package, used as a build_requirement that contains some `cmake` modules. When using 2 profiles, `CMAKE_PREFIX_PATH` from build_reqs is not being exposed on any cmake generator. Would your proposal address this?

**Comment by blackliner** [comment]
Thanks for the insights @lasote 

In my case, for capnproto, It would be ok to find the host version both times, because all I need during the build context are some cmake macros and they dont care about platforms 😝

**Comment by lasote** [comment]
@ericriff Yes, it should be declared when doing the `find_package(my_internal_packageBuild)`.
The issue in general with the `CMAKE_PREFIX_PATH` using `protobuf` as an example, required as build_require and as a regular require at the same time, and also with the multi-config packages, is that the only thing we can do is append all the `CMAKE_PREFIX_PATH` for all the configurations and build/host in the same variable because it is not something we can assign to a target, the same issue that we have with the `build_modules`.

**Comment by ericriff** [comment]
Will the variables in the build module also have a different name?
As an example now the `protobuf` package provides `Protobuf_PROTOC_EXECUTABLE` which points to the `protoc` exec.
If I have a CMakeLists.txt where I need to use protoc to generate files and then link against protobuf libraries I would have to do

```
find_package(Protobuf) # For the libs, lets say arm
find_package(ProtobufBuild) # For Protoc, lets say Linux x86_64
```

Then bothe modules will provide `Protobuf_PROTOC_EXECUTABLE`? 
Even if the autogenerated variables have different names, there are a bunch of recipes that generates extra CMake variables to have compatibility with the official find module that the library provides. Those packages, like protobuf, I believe will have colliding variables.

**Comment by jgsogo** [comment]
The current `protobuf` recipe in ConanCenter manages this situation in a very easy way (kudos to the community). Any other code generation tool with the same approach as protocol buffers could implement the same in the recipe and it will work. 

As it is now, there is no need to modify existing `CMakeLists.txt` in projects that are using `find_package(Protobuf)`, they will work out of the box, also in a cross-compiling scenario, no need to modify the build-system files to add an extra `find_package`.

**Comment by blackliner** [comment]
Could you elaborate on the "very easy way"? Maybe instead of changing conan, if there are designs available, why not come up with new recommendations for recipes? And maybe have some helper methods from conan?

**Comment by jgsogo** [comment]
Recipe uses the following in the generated `.cmake` files:

```cmake
            if(NOT TARGET protobuf::protoc)
                if(CMAKE_CROSSCOMPILING)
                    find_program(PROTOC_PROGRAM protoc PATHS ENV PATH NO_DEFAULT_PATH)
                endif()
                if(NOT PROTOC_PROGRAM)
                    set(PROTOC_PROGRAM \"${{CMAKE_CURRENT_LIST_DIR}}/{protoc_rel_path}\")
                endif()
                get_filename_component(PROTOC_PROGRAM \"${{PROTOC_PROGRAM}}\" ABSOLUTE)
                set(Protobuf_PROTOC_EXECUTABLE ${{PROTOC_PROGRAM}} CACHE FILEPATH \"The protoc compiler\")
                add_executable(protobuf::protoc IMPORTED)
                set_property(TARGET protobuf::protoc PROPERTY IMPORTED_LOCATION ${{Protobuf_PROTOC_EXECUTABLE}})
            endif()
```

Depending on the scenario (cross-compiling or not), the variable `Protobuf_PROTOC_EXECUTABLE` and the target `protobuf::protoc` will use the one from _host_ or the one from _build_ context. Basically, it takes advantage of the environment being populated by Conan with the PATH to the tools provided by build-requires.

**Comment by theodelrieu** [comment]
@lasote 

Hi, I think it's better than not solving the issue for sure. Is there something preventing generators other than `CMakeDeps` to have this functionality?

**Comment by lasote** [comment]
> Hi, I think it's better than not solving the issue for sure.

Do you have a better idea? Ideas are welcome! 

> Is there something preventing generators other than CMakeDeps to have this functionality?

Not technical reasons, but Conan 2.0 will have only the CMakeDeps generator and we consider we have to stop at some point evolving the old generators to focus on the goal.

**Comment by theodelrieu** [comment]
>Do you have a better idea? Ideas are welcome!

Nope! I've let dual profiles on the backburner for a few months so I won't have an enlightenment any time soon 😆.

 >we consider we have to stop at some point evolving the old generators to focus on the goal.
 
 Cannot agree more.

**Comment by ericriff** [comment]
Interesting that Conan 2.0 will drop all generators but CMakeDesps. Is there a reason for that? Maybe a blog entry explaining it?
Also, is there a ETA for Conan 2.0? (So I can prepare our codebase in advance)

**Comment by memsharded** [comment]
There is no  ETA for Conan 2.0, it will take some time for sure, not before the end of the year. Discussions happen in the https://conan.io/tribe.html git repo https://github.com/conan-io/tribe

Reasons:
- We need to evolve the generators, to account for new models, like this one (build/host profiles). We can't do this in the previous generators without breaking and this is really blocking capability of evolution in the others.
- CMakeDeps is a fork of ``cmake_find_package_multi`` which in turn is an evolution of ``cmake_find_package``. So CMakeDeps supersed both of them.
- Also, it is very confusing for users to have multiple different generators, with different capabilities. The basic ``cmake`` and ``cmake_multi`` generators have been heavily criticized by the community for being intrusive and not the cmake way, so ``find_package()`` is the way to go moving forward, and this is the reason those will be removed.

**Comment by SeanSnyders** [comment]
We are blocked with cross-compilation via NDK package `android-ndk/r22` and using `cmake_find_package` generators for our own and external libraries. These libraries are either pure binary lib packages and/or "toolchain" packages.

Would the `CMakeDeps` generator solve this? Will it solve it without having *all* packages update their CMake projects in some way? It would not help us at all if we need to wait for all our transitive dependencies to update something to align with this use-case until we can use it in our projects.
And how usable is the current _experimental_ `CMakeDeps` generator in `Conan v1` to how it will be in `Conan v2`?

I also support the suggestion of writing a blog post that will show the end-form of how a project will look like using `CMakeDeps` and how e.g. it would work with GTest, or cross-compilation workflows. Things are very unclear at the moment, and for us an enterprise-wide adoption of Conan is looking risky.

**Comment by memsharded** [comment]
Hi @SeanSnyders 

> Would the CMakeDeps generator solve this? Will it solve it without having all packages update their CMake projects in some way? It would not help us at all if we need to wait for all our transitive dependencies to update something to align with this use-case until we can use it in our projects.

Yes, that is the idea, CMakeDeps should solve this. No, all projects will not need to migrate, thanks to ``package_info()`` this should be decoupled, and you should only use ``CMakeDeps`` in those projects that are blocked now and you need this new functionality, while the upstream packages can still be built with other generators.

>And how usable is the current experimental CMakeDeps generator in Conan v1 to how it will be in Conan v2?

Yes, at the moment CMakeDeps is a bit experimental, not ready for production, but we are heavily investing efforts in it, to solve this issue as soon as possible. The idea is that it will become much more stable well in advance of Conan 2.0, to allow users to start using it for a smooth transition to 2.0.

I would recommend to wait 1 or 2 releases to start using CMakeDeps, we will certainly write a blog post clearly explaining it.

**Comment by SeanSnyders** [comment]
Having an issue with generating the required files via `CMakeDeps` generator in the most basic configuration of that generator using v1.37 where the generator worked fine in previous versions of Conan. Logged an issue: https://github.com/conan-io/conan/issues/9047

**Comment by blackliner** [comment]
So, for my initial issue, in a cross compilation situation, there is no way to get the `FindCapnProto.cmake` if it is a build_requirement only?

The library `CapnProto` cannot be compiled on the host platform (QNX), so adding it to both build_requirements and requirements is not possible.

`CMakeDeps` does not provide any reasonable files other than a `cmakedeps_macros.cmake` with zero references to CapnProto

@jgsogo you mentioned this above:
```
Conan will populate the environment before entering build() method with the information provided in the env_info of the capnproto recipe.
```
But that only relates to optional ENV settings? How could a CMakeLists.txt find the package mentioned in build_requirements?

**Comment by memsharded** [comment]
Hi @blackliner 

> So, for my initial issue, in a cross compilation situation, there is no way to get the FindCapnProto.cmake if it is a build_requirement only?

Yes, there is, the ways is documented in (conan 1.37) https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#build-context-activated, it should work for a ``build_requires`` only too. Have you tried that and it didn't work? In that case, please open a new ticket, it would be considered a bug and be fixed.

**Comment by blackliner** [comment]
Oh this is awesome, will try it out ASAP. But this is only `CMakeDeps` I guess 🤔 Need it for `cmake_find_package`

**Comment by memsharded** [comment]
As a hint of the Conan dev process:

- When an issue like this one is closed automatically by a PR, it has been implemented or fixed by that PR.
- Both the PR and the issue should have the milestone that matches the Conan release it belongs.
- The first comment of the PR contains also a link to the documentation PR, for example https://github.com/conan-io/conan/pull/8985, that fixes this issue, contains a link to https://github.com/conan-io/docs/pull/2104/files, that makes explicit what is changed in the docs.

**Comment by memsharded** [comment]
> Need it for cmake_find_package

It is not possible to introduce this for ``cmake_find_package`` as it would be breaking

**Comment by blackliner** [comment]
Got that, and the CMakeDeps looks good, but I am missing build_modules. They should be part of the generated files, I see for example:
`CapnProtoConfig.cmake`
```
# Only the first installed configuration is included to avoid the collission
foreach(_BUILD_MODULE ${capnproto_BUILD_MODULES_PATHS_RELEASE} )
    conan_message(STATUS "Conan: Including build module from '${_BUILD_MODULE}'")
    include(${_BUILD_MODULE})
endforeach()
```
and
`conan/recipes/resim/build/CapnProto-release-x86_64-data.cmake`
```
set(capnproto_BUILD_MODULES_PATHS_RELEASE "${capnproto_PACKAGE_FOLDER_RELEASE}/lib/cmake/CapnProto/CapnProtoMacros.cmake")
```
but still get:
```
  Unknown CMake command "capnp_generate_cpp".
```

Known issue, shall I try to debug a bit more, or time for a ticket?

**Comment by blackliner** [comment]
Ok, scratch that, I was looking at the wrong folder. The capnproto_BUILD_MODULES_PATHS_RELEASE is actually empty!

Found it: build_context_build_modules  https://github.com/conan-io/conan/issues/7719#issuecomment-848846803

Case closed!

**Comment by zuowanbushiwo** [comment]
@jgsogo @lasote @madebr @SeanSnyders @ericriff 
I have been stuck with the problem of cross-compiling between protobuf and *..proto for two days. I have read a lot of issues and documents, but have not found a complete solution. What should I do? Where can I find a example?  Thanks!

**Comment by madebr** [comment]
@zuowanbushiwo 
Have you taken a look [at the test package of protobuf](https://github.com/conan-io/conan-center-index/tree/master/recipes/protobuf/all/test_package)?

I just tried to [build protobuf's tutorial](https://developers.google.com/protocol-buffers/docs/cpptutorial) using conan + protobuf.
I've [created a gist](https://gist.github.com/madebr/805efee409211d042a57d7fa6f06c4ba).

I can build it natively using my default profile, or cross build to Windows using mingw (using `-pr:h ... -pr:b ...` syntax.
I use `conan install .. -pr:h .. -pr:b ..` and `conan build ...` to ensure that the environment variables are set correctly.

When using `virtualenv`, it only works when you explicitly set `CMAKE_BUILD_TYPE`.
This is a gotcha that I just discovered (probably due to my usage of `cmake_find_package_multi`).

**Comment by zuowanbushiwo** [comment]
@jgsogo @lasote @madebr @SeanSnyders @ericriff

Thank you very much for your help. I tried both repositories(conan-center-index/recipes/protobuf and your gist.) and both failed to compile and the errors were the same.
1. I follow conan doc  [Linux/Windows/macOS to Android ](https://docs.conan.io/en/latest/systems_cross_building/cross_building.html#linux-windows-macos-to-android), create a android profile host， as follow：
```

include(default)

target_host=aarch64-linux-android
api_level=24

[settings]
arch=armv8
build_type=Release
compiler=clang
compiler.libcxx=libc++
compiler.version=11
os=Android
os.api_level=$api_level

[build_requires]

[options]

[conf]
tools.android:ndk_path=/opt/android-sdk/ndk-bundle

[env]
CONAN_CMAKE_TOOLCHAIN_FILE=/opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake
PATH=[/opt/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin]
CHOST=$target_host
AR=$target_host-ar
AS=$target_host-as
RANLIB=$target_host-ranlib
CC=$target_host$api_level-clang
CXX=$target_host$api_level-clang++
LD=$target_host-ld
STRIP=$target_host-strip
```

2.  for conan-center-index/recipes/protobuf , I use the following command  follow  document guide:  
`conan-center-index/recipes/protobuf/all$ conan create . 3.15.5@ -pr:h  ~/conan_exsample/conan-center-index/recipes/android  -pr:b default  --build`
build depends protobuf and zib success， build test package， A major error occurred：
```

CMake Error at CMakeLists.txt:7 (find_package):
  Could not find a package configuration file provided by "protobuf" with any
  of the following names:

    protobufConfig.cmake
    protobuf-config.cmake

  Add the installation prefix of "protobuf" to CMAKE_PREFIX_PATH or set
  "protobuf_DIR" to a directory containing one of the above files.  If
  "protobuf" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package/build/231d0c73cb5a9d0b3a61fdaeff116c5c0990ae0e/CMakeFiles/CMakeOutput.log".
ERROR: protobuf/3.15.5 (test package): Error in build() method, line 16
	cmake.configure()
	ConanException: Error 1 while executing cd '/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package/build/231d0c73cb5a9d0b3a61fdaeff116c5c0990ae0e' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_TOOLCHAIN_FILE="/opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="clang" -DCONAN_COMPILER_VERSION="11" -DCONAN_LIBCXX="libc++" -DCMAKE_INSTALL_PREFIX="/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package/build/231d0c73cb5a9d0b3a61fdaeff116c5c0990ae0e/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_PREFIX_PATH="/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package/build/231d0c73cb5a9d0b3a61fdaeff116c5c0990ae0e" -DCMAKE_MODULE_PATH="/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package/build/231d0c73cb5a9d0b3a61fdaeff116c5c0990ae0e" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -Dprotobuf_LITE="False" -Wno-dev '/home/packager/conan_exsample/conan-center-index/recipes/protobuf/all/test_package
```
3. build your gist，i  Use the following command：
 First   ` conan install --install-folder conan-build-cross  --profile:host ~/conan_exsample/conan-center-index/recipes/android  --build outdated   --profile:build default  .`
this step may successful ,and The contents of the conan-build-cross directory are as follows:
![image](https://user-images.githubusercontent.com/3354428/131236150-76b59038-93be-4817-87ad-b4deb1a6c6e6.png)

Second:  `conan build --build-folder conan-build-cross  .`  have a error output，same as the first case above:
<details>
<summary>**ALL ERROR  LOG**</summary>
<pre><code>

```

Using lockfile: '/home/packager/conan_exsample/proto_gist/conan-build-cross/conan.lock'
Using cached profile from lockfile
conanfile.py: Calling build()
-- ANDROID_PLATFORM not set. Defaulting to minimum supported version
16.
CMake Warning at /opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake:449 (message):
  An old version of CMake is being used that cannot automatically detect
  compiler attributes.  Compiler identification is being bypassed.  Some
  values may be wrong or missing.  Update to CMake 3.19 or newer to use
  CMake's built-in compiler identification.
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/CMakeDetermineSystem.cmake:93 (include)
  CMakeLists.txt:2 (project)


-- Check for working CXX compiler: /opt/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++
CMake Warning at /opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake:449 (message):
  An old version of CMake is being used that cannot automatically detect
  compiler attributes.  Compiler identification is being bypassed.  Some
  values may be wrong or missing.  Update to CMake 3.19 or newer to use
  CMake's built-in compiler identification.
Call Stack (most recent call first):
  /home/packager/conan_exsample/proto_gist/conan-build-cross/CMakeFiles/3.16.3/CMakeSystem.cmake:6 (include)
  /home/packager/conan_exsample/proto_gist/conan-build-cross/CMakeFiles/CMakeTmp/CMakeLists.txt:3 (project)


-- Check for working CXX compiler: /opt/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -- works
-- Detecting CXX compiler ABI info
CMake Warning at /opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake:449 (message):
  An old version of CMake is being used that cannot automatically detect
  compiler attributes.  Compiler identification is being bypassed.  Some
  values may be wrong or missing.  Update to CMake 3.19 or newer to use
  CMake's built-in compiler identification.
Call Stack (most recent call first):
  /home/packager/conan_exsample/proto_gist/conan-build-cross/CMakeFiles/3.16.3/CMakeSystem.cmake:6 (include)
  /home/packager/conan_exsample/proto_gist/conan-build-cross/CMakeFiles/CMakeTmp/CMakeLists.txt:3 (project)


-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: Adjusting output directories
-- Conan: Using cmake targets configuration
-- Library protoc found /home/packager/.conan/data/protobuf/3.17.1/_/_/package/d6929ec3b48db040b8a53bc090789ebcf2df8913/lib/libprotoc.a
-- Library protobuf found /home/packager/.conan/data/protobuf/3.17.1/_/_/package/d6929ec3b48db040b8a53bc090789ebcf2df8913/lib/libprotobuf.a
-- Library z found /home/packager/.conan/data/zlib/1.2.11/_/_/package/2b2f1542da8962686483ada8f65b191b8b9dc9da/lib/libz.a
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan: Compiler Clang>=8, checking major version 11
-- Conan: Checking correct version: 11
-- Conan: C++ stdlib: libc++
CMake Error at CMakeLists.txt:9 (find_package):
  Could not find a package configuration file provided by "protobuf" with any
  of the following names:

    protobufConfig.cmake
    protobuf-config.cmake

  Add the installation prefix of "protobuf" to CMAKE_PREFIX_PATH or set
  "protobuf_DIR" to a directory containing one of the above files.  If
  "protobuf" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "/home/packager/conan_exsample/proto_gist/conan-build-cross/CMakeFiles/CMakeOutput.log".
ERROR: conanfile.py: Error in build() method, line 18
	cmake.configure()
	ConanException: Error 1 while executing cd '/home/packager/conan_exsample/proto_gist/conan-build-cross' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_TOOLCHAIN_FILE="/opt/android-sdk/ndk-bundle/build/cmake/android.toolchain.cmake" -DCONAN_IN_LOCAL_CACHE="OFF" -DCONAN_COMPILER="clang" -DCONAN_COMPILER_VERSION="11" -DCONAN_LIBCXX="libc++" -DCMAKE_INSTALL_PREFIX="/home/packager/conan_exsample/proto_gist/conan-build-cross/package" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCMAKE_PREFIX_PATH="/home/packager/conan_exsample/proto_gist/conan-build-cross" -DCMAKE_MODULE_PATH="/home/packager/conan_exsample/proto_gist/conan-build-cross" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -DCMAKE_VERBOSE_MAKEFILE="ON" -Wno-dev '/home/packager/conan_exsample/proto_gist'

```
</code></pre>
</details>

4.  conan-build-cross  build output directory files
[conanbuildinfo.cmake.txt](https://github.com/conan-io/conan/files/7071809/conanbuildinfo.cmake.txt)
[conanbuildinfo.txt](https://github.com/conan-io/conan/files/7071810/conanbuildinfo.txt)
[conaninfo.txt](https://github.com/conan-io/conan/files/7071811/conaninfo.txt)
[graph_info.json.txt](https://github.com/conan-io/conan/files/7071812/graph_info.json.txt)

5. I used a android docker for test，[here](https://github.com/nephatrine/docker-nxbuilder/blob/android/Dockerfile) is dockerfile， [here ](https://hub.docker.com/layers/nephatrine/nxbuilder/android/images/sha256-d13f4af8e8a5ee1be6f57566648339f63ecd05413f10a0125dff0eea06de8a1a?context=explore)is docker images。

thanks！
best regards

**Comment by blackliner** [comment]
> Depending on the scenario (cross-compiling or not), the variable `Protobuf_PROTOC_EXECUTABLE` and the target `protobuf::protoc` will use the one from _host_ or the one from _build_ context. Basically, it takes advantage of the environment being populated by Conan with the PATH to the tools provided by build-requires.

@jgsogo holy smokes, took me some strokes to figure that out... So basically it works because this line:

https://github.com/conan-io/conan-center-index/blob/master/recipes/protobuf/all/conanfile.py#L235
```
self.env_info.PATH.append(bindir)
```

is called **FIRST** for the `build_requires`, so it **APPENDS** the `bindir` to the **PATH** env. After that, the "normal" `requires` also **APPENDS** to **PATH**, but since `find_program(PROTOC_PROGRAM protoc PATHS ENV PATH NO_DEFAULT_PATH)` is looking from "left to right", it finds the `build_requirements` version first? 🤯

**Comment by blackliner** [comment]
@anton-danielsson 's work? nice!

**Comment by anton-danielsson** [comment]
@blackliner I have also had some pretty intense wrestling matches with CMake, Conan, Protobuf and cross-compiling.
Don't mention GRPC... :D

But yes, it's the line you mention together with the fact that `find_program` in find protobuf prefers the path (`PATHS ENV PATH NO_DEFAULT_PATH`).
https://github.com/conan-io/conan-center-index/blob/master/recipes/protobuf/all/conanfile.py#L127

I'm not sure how robust this actually is though...
Seems safe to depend on the "looking left to right" fact.
Not sure it's actually safe to depend on the order Conan appends to the path though.

**Comment by ericriff** [comment]
I'm a bit late to the party, but this is how we're calling protoc it from the CMake side
```
add_custom_command(
    OUTPUT ${proto_generated_srcs}
    COMMAND
        ${Protobuf_PROTOC_EXECUTABLE}
        --cpp_out=${CMAKE_CURRENT_BINARY_DIR}
        --proto_path=${CMAKE_CURRENT_LIST_DIR} ${PROTO_FILES}
    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
    DEPENDS ${PROTO_FILES} ${Protobuf_PROTOC_EXECUTABLE}
    COMMENT "Generating protobuf files for ${PROTO_FILES}"
    VERBATIM
)
```
`Protobuf_PROTOC_EXECUTABLE` is one of the CMake variables that the official `findProtobuf.cmake` provides and has been added to the recipe in an update a few months ago.

**Comment by SeanSnyders** [comment]
Any movement on this in general? There is a lot of technical in-depth scenarios which makes it hard to follow whether progress is being made on a higher level.
The overarching topic of cross-compiling and use through CMakeDeps and host and build profiles. That mentioned blog post and/or usage documentation would be pretty great about now.

**Comment by lasote** [comment]
https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html

We have there some docs about the implementation of the CMakeDeps.
